src/kokkos_abstraction.hpp

```diff
@@ -48,8 +48,14 @@ using HostExecSpace = Kokkos::DefaultHostExecutionSpace;
 using LayoutWrapper = Kokkos::LayoutRight;
 using MemUnmanaged = Kokkos::MemoryTraits<Kokkos::Unmanaged>;
 
-#if defined(KOKKOS_ENABLE_CUDA) && defined(PARTHENON_ENABLE_HOST_COMM_BUFFERS)
+#if defined(PARTHENON_ENABLE_HOST_COMM_BUFFERS)
+#if defined(KOKKOS_ENABLE_CUDA)
 using BufMemSpace = Kokkos::CudaHostPinnedSpace::memory_space;
+#elif defined(KOKKOS_ENABLE_HIP)
+using BufMemSpace = Kokkos::Experimental::HipHostPinnedSpace::memory_space;
+#else
+#error "Unknow comm buffer space for chose execution space."
```

Comment:
I tried to keep the logic simple and make it crash hard for unintended/unsupported cases.
For example, I'm not sure about the status of the memory space equivalent for the SYCL backend.
And for OpenMP it would depend on whether this is OpenMP or OpenMPTarget with the former not having any meaning and the latter I'm not sure about (again if it's supported etc).
As far as my testing goes "hostpinnedmemory" should only be used as an exception or for testing because in practice it was always slower.
That's why I didn't set a default but rather make it crash.
You prefer a cmake solution?

Reply:
> Is there a reason to bother erroring rather than just setting BufMemSpace to the default?

I am not in favor of `HOST_COMM_BUFFERS` quietly doing nothing if you set it with an openmp backend. On top of that, is it always true that openmp backend will do nothing? If we ever try to use openmp target offload, is there a host-pinned equivalent?

> You prefer a cmake solution?

Even if we go for a cmake solution, I would not remove an error message from `kokkos_abstraction`. We could make the error more informative for cases where incompatible macros have been chosen though? That could be done with the preprocessor or cmake... or with `static_assert(false, "message")`

Reply:
I'm not that concerned about how exactly things fail when trying to compile or SYCL or another unsupported case, I think anyone attempting those things knows what they're doing enough that specific errors don't matter.

Instead, I'm imagining the case where someone is looking down the Parthenon options wishing to compile for OpenMP/CPU, finds a "HOST_BUFFERS" option without any other warnings or prefixes, and turns it on assuming that buffers, like all other computations in this case, should be on the host.  Or, someone repurposes a file for compiling against their desktop GPU one day to compile for CPU, and assumes that they can just change the Kokkos_ENABLE_X options.  Flooding the scrollback with errors seems hostile in these cases, vs. just putting the buffers on the host.  I'm also just generally not a fan of producing errors at compile time when they could have been caught on configure.

If it's important to keep the error in unsupported contexts (can we just check for those in the CMakeLists too?), we should document that the option is for use only in limited contexts:
```cmake
option(PARTHENON_ENABLE_HOST_COMM_BUFFERS "CUDA/HIP Only: Allocate communication buffers on host (slow)" OFF)
```

We can even just check it against the relevant flags:
```cmake
if (PARTHENON_ENABLE_HOST_COMM_BUFFERS)
  if (NOT Kokkos_ENABLE_CUDA AND NOT Kokkos_ENABLE_HIP)
    message(FATAL_ERROR "Host pinned buffers for MPI communication are supported only for CUDA and HIP backends.)
  endif()
endif()
```

Reply:
Yeah I'm fine with an informative error message in `cmake`, if that reduces the C++ error overload. My point is just that we *should* have an error in the preprocessor for any cases we missed. Just to be a bit more defensive.

Reply:
I like both suggested solutions and will update accordingly.

PS: As Jonah mentioned, so far we've typically chosen the "fail loud" approach for anything that we could reasonable catch as unsupported and untested. Too many got burned in the past by a code that silently ran in that regime.

Reply:
I agree with leaving the #error (with a couple more 'n's), if it has a guard in the CMake file; my issue was that without a clear error message it could easily become a sharp edge for new users, but I think a configure-time error makes it pretty foolproof.

Separately, though, it seems obvious to me that in the context of compiling for OpenMP on the host, asking for HOST_COMM_BUFFERS should be a no-op, because all the buffers are on the host anyway.  If others don't agree then we can keep it totally explicit, but "are we compiling for CPU" seems like something simple enough to determine during the configuration step (or which we should be determining anyway...)

Reply:
To me at least it's not obvious that `HOST_COMM_BUFFERS` should be a no-op with `OpenMP` on host. I'd rather the `cmake` explain what's going on rather than let that quietly do nothing.

Reply:
Huh, okay, thanks.  Will update my prior then.  Resolving this, to expedite the merge